ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/debug/BreakpointStorageImpl.java

```diff
@@ -0,0 +1,260 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.logging.Logger;
+
+import static org.eclipse.che.ide.api.debug.Breakpoint.Type.BREAKPOINT;
+
+/**
+ * Breakpoints storage based on local storage.
+ *
+ * @author Anatolii Bazko
+ */
+@Singleton
+public class BreakpointStorageImpl implements BreakpointStorage {
+
+    private static final Logger LOG                           = Logger.getLogger(BreakpointStorageImpl.class.getName());
+    private static final String LOCAL_STORAGE_BREAKPOINTS_KEY = "che-breakpoints-";
+
+    private final AppContext             appContext;
+    private final DtoFactory             dtoFactory;
+    private final Storage                storage;
+    private final WorkspaceServiceClient workspaceServiceClient;
+    private final JsPromiseProvider      promiseProvider;
+    private final Promise<Void>          readAllBreakpointMarker;
+    private final List<Breakpoint>       breakpoints;
+    private final EventBus               eventBus;
+
+    @Inject
+    public BreakpointStorageImpl(AppContext appContext,
+                                 DtoFactory dtoFactory,
+                                 WorkspaceServiceClient workspaceServiceClient,
+                                 JsPromiseProvider promiseProvider, EventBus eventBus) {
+        this.appContext = appContext;
+        this.dtoFactory = dtoFactory;
+        this.workspaceServiceClient = workspaceServiceClient;
+        this.promiseProvider = promiseProvider;
+        this.eventBus = eventBus;
+        this.storage = Storage.getLocalStorageIfSupported();
+        this.breakpoints = new LinkedList<>();
+
+        if (storage == null) {
+            LOG.warning("Local storage is not supported. Breakpoints won't be preserved.");
+            this.readAllBreakpointMarker = promiseProvider.resolve(null);
+        } else {
+            this.readAllBreakpointMarker = getReadAllBreakpointMarker();
+            preserve();
+            clearOutdatedRecords();
+        }
+    }
+
+    @Override
+    public void addAll(final List<Breakpoint> breakpoints) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.addAll(breakpoints);
+
+            if (storage != null) {
+                preserve();
+            }
+        });
+    }
+
+    @Override
+    public void add(final Breakpoint breakpoint) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.add(breakpoint);
+
+            if (storage != null) {
+                preserve();
+            }
+        });
+    }
+
+    @Override
+    public void delete(final Breakpoint breakpoint) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.removeIf(
+                    b -> b.getLineNumber() == breakpoint.getLineNumber() && b.getPath().equals(breakpoint.getPath())
+            );
+
+            if (storage != null) {
```

Comment:
You can place one check for nullable storage into method `preserve` instead of doing this many times.